module-system/sov-modules-api/src/default_signature.rs

```
@@ -2,70 +2,88 @@
 use std::str::FromStr;
 
 use borsh::{BorshDeserialize, BorshSerialize};
-use ed25519_dalek::ed25519::signature::Signature as DalekSignatureTrait;
 use ed25519_dalek::{
-    PublicKey as DalekPublicKey, Signature as DalekSignature, PUBLIC_KEY_LENGTH, SIGNATURE_LENGTH,
+    Signature as DalekSignature, VerifyingKey as DalekPublicKey, PUBLIC_KEY_LENGTH,
 };
-use serde::de::Error;
-use serde::{Deserialize, Deserializer, Serialize, Serializer};
 
 use crate::{SigVerificationError, Signature};
 
 #[cfg(feature = "native")]
 pub mod private_key {
-
-    use ed25519_dalek::{Keypair, SignatureError, Signer};
+    use ed25519_dalek::{Signer, SigningKey, KEYPAIR_LENGTH, SECRET_KEY_LENGTH};
     use rand::rngs::OsRng;
     use thiserror::Error;
 
     use super::{DefaultPublicKey, DefaultSignature};
     use crate::{Address, PrivateKey, PublicKey};
 
     #[derive(Error, Debug)]
-    pub enum DefaultPrivateKeyHexDeserializationError {
+    pub enum DefaultPrivateKeyDeserializationError {
         #[error("Hex deserialization error")]
         FromHexError(#[from] hex::FromHexError),
-        #[error("PrivateKey deserialization error")]
-        PrivateKeyError(#[from] SignatureError),
+        #[error("KeyPairError deserialization error")]
+        KeyPairError(#[from] ed25519_dalek::SignatureError),
+        #[error("Invalid private key length: {actual}, expected {expected_1} or {expected_2}")]
+        InvalidPrivateKeyLength {
+            expected_1: usize,
+            expected_2: usize,
+            actual: usize,
+        },
+        // This error is thrown only if private methods are misused.
+        #[error("Potential bug: failed to convert from slice")]
+        SliceError(#[from] std::array::TryFromSliceError),
```

I think these should be a panic via `expect` instead of returning an error, what do you think? If this ever happens, it means a serious invariant was broken in our code and we must work on a patch.

---

Agree!

---

Nice, I guess we can also remove the invariant altogether :100:

---

Done!